proxysrv/server.go

```diff
@@ -0,0 +1,382 @@
+// Package proxysrv proxies metrics over gRPC to global veneurs using
+// consistent hashing
+//
+// The Server provided accepts a hash ring of destinations, and then listens
+// for metrics over gRPC.  It hashes each metric to a specific destination,
+// and forwards each metric to its appropriate destination Veneur.
+package proxysrv
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context" // This can be replace with "context" after Go 1.8 support is dropped
+	"google.golang.org/grpc"
+	"stathat.com/c/consistent"
+
+	"github.com/stripe/veneur/forwardrpc"
+	"github.com/stripe/veneur/samplers"
+	"github.com/stripe/veneur/samplers/metricpb"
+	"github.com/stripe/veneur/ssf"
+	"github.com/stripe/veneur/trace"
+	"github.com/stripe/veneur/trace/metrics"
+)
+
+const (
+	// By default, cancel forwarding operations after 10 seconds
+	defaultForwardTimeout = 10 * time.Second
+
+	// Report server statistics every 10 seconds
+	reportStatsInterval = 10 * time.Second
+)
+
+// Server is a gRPC server that implements the forwardrpc.Forward service.
+// It receives metrics and forwards them consistently to a destination, based
+// on the metric name, type and tags.
+type Server struct {
+	*grpc.Server
+	destinations *consistent.Consistent
+	opts         *options
+	conns        *clientConnMap
+	updateMtx    sync.Mutex
+
+	// A simple counter to track the number of goroutines spawned to handle
+	// proxying metrics
+	activeProxyHandlers *int64
+}
+
+// Option modifies an internal options type.
+type Option func(*options)
+
+type options struct {
+	log            *logrus.Entry
+	forwardTimeout time.Duration
+	traceClient    *trace.Client
+}
+
+// New creates a new Server with the provided destinations. The server returned
+// is unstarted.
+func New(destinations *consistent.Consistent, opts ...Option) (*Server, error) {
+	res := &Server{
+		Server: grpc.NewServer(),
+		opts: &options{
+			forwardTimeout: defaultForwardTimeout,
+		},
+		conns:               newClientConnMap(grpc.WithInsecure()),
+		activeProxyHandlers: new(int64),
+	}
+
+	for _, opt := range opts {
+		opt(res.opts)
+	}
+
+	if res.opts.log == nil {
+		log := logrus.New()
+		log.Out = ioutil.Discard
+		res.opts.log = logrus.NewEntry(log)
+	}
+
+	if err := res.SetDestinations(destinations); err != nil {
+		return nil, fmt.Errorf("failed to set the destinations: %v", err)
+	}
+
+	forwardrpc.RegisterForwardServer(res.Server, res)
+
+	return res, nil
+}
+
+// Serve starts a gRPC listener on the specified address and blocks while
+// listening for requests. If listening is interrupted by some means other than
+// Stop or GracefulStop being called, it returns a non-nil error.
+func (s *Server) Serve(addr string) (err error) {
+	ln, err := net.Listen("tcp", addr)
```

Comment:
This is a bit of a nit, but as long as we're doing things with defer-assignment of return values, i'd rather be as clear about variable scoping as possible.

Either we should use `:=` here, and _not_ use a named return value (so that this line initializes `err`), or we use the named return value and then add a `var ln net.Listener` declaration the line before, and assign with `=`. i'd prefer the former.

Reply:
I think I'm definitely convinced at this point that trying to handle errors in deferred calls like this is just tricky to do right, and I'm not really sure it gains us much in this case specifically.

I went with your suggestion, and just removed the named return value and the extra error handling when closing the listener.